RecoTauTag/RecoTau/plugins/TauDiscriminationAgainstCaloMuon.cc

```diff
@@ -280,8 +270,7 @@ double TauDiscriminationAgainstCaloMuon<TauType, TauDiscriminator>::discriminate
 
   const CaloSubdetectorGeometry* ebGeometry = caloGeometry_->getSubdetectorGeometry(DetId::Ecal, EcalBarrel);
   const CaloSubdetectorGeometry* eeGeometry = caloGeometry_->getSubdetectorGeometry(DetId::Ecal, EcalEndcap);
-  const CaloSubdetectorGeometry* hbGeometry = caloGeometry_->getSubdetectorGeometry(DetId::Hcal, HcalBarrel);
-  const CaloSubdetectorGeometry* heGeometry = caloGeometry_->getSubdetectorGeometry(DetId::Hcal, HcalEndcap);
+  const HcalGeometry* hcGeometry = (HcalGeometry*)(caloGeometry_->getSubdetectorGeometry(DetId::Hcal, HcalBarrel));
```

Comment:
wouldn't it make more sense to change the interface of CaloGeometry to add ```getSubdetectorGeometry(det)``` so that we do not have this counter-intuitive code which semantically looks like HcalBarrel subdetector is selected, while in reality it's all of Hcal.

Reply:
i agree with @slava77 comment - this is a very confusing (and hacky) function call.. how can this be cleaned up?

Reply:
I shall think about it. It is not a very easy solution - there are detectors which has one geometry for all sub detectors while there are detectors which have different geometry for each sub detector.  I shall try to see if something can be done in near future. But that has to be a separate PR and this PR is simply a bug fix

Reply:
it would seem a simple solution is to add a getDetectorGeometry and implement getSubdetectorGeometry only when relevant (throwing an exception otherwise). That way you don't have to find these "bugs" (rather, code that didn't follow an interface change that didn't change the interface itself) one by one

Reply:
David EcalBarrelGeometry, HcalGeometry, .. all are derived from CaloSubDetectorGeometry and depending on the detector/subdetector the functionalities are somewhat different. I do not think what you are suggesting makes sense